### DISCUSSION TITLE
ESP32-S3 NEC code not sent correctly

### DISCUSSION THREAD

**Discussion by cytan299** [Arduino-IRremote/Arduino-IRremote#1169] [Help]
HI,
  Thanks for writing this library. It's greatly appreciated.

  Unfortunately,  I'm having problems sending IR codes with the ESP32-s3. (It's probably pilot error).

   First I used ReceiveDemo.ino to get the IR code for vol+ for my Vizio TV (note: I used D12 for receive)

********
Protocol=NEC Address=0x4 Command=0x2 Raw-Data=0xFD02FB04 32 bits LSB first
Send with: IrSender.sendNEC(0x4, 0x2, \<numberOfRepeats\>);
*********

Then for sending, I did the following in my ESP32 (I disabled receive and used D12 for transmit)
***********
#define DISABLE_CODE_FOR_RECEIVER 
#define NO_LED_FEEDBACK_CODE
#define IR_SEND_PIN	12
#include <IRremote.hpp>

void setup(){
...
  IrSender.begin(DISABLE_LED_FEEDBACK); 
....
}

void doit()
{
 ...
   IrSender.sendNEC(0x4,0x2,3);
...
}

*********
Unfortunately the above code didn't change the volume of my TV. So, I programmed **another** ESP32 to read the IR with RecevieDemo.ino. I got the following result from the above ESP32:

*********
20:01:18.889 -> 
20:01:18.889 -> Protocol=Onkyo Address=0xFBFC Command=0xFDFA Raw-Data=0xFDFAFBFC 32 bits LSB first
20:01:18.889 -> Send with: IrSender.sendOnkyo(0xFBFC, 0xFDFA, <numberOfRepeats>);
20:01:18.939 -> 
20:01:18.939 -> Protocol=UNKNOWN Hash=0x0 2 bits (incl. gap and start) received
20:01:18.939 -> Received noise or an unknown (or not yet enabled) protocol
20:01:18.939 -> rawData[4]: 
20:01:18.939 ->  -23400
20:01:18.939 ->  +9000,-3000
20:01:18.939 ->  + 600
20:01:18.939 -> Sum: 12600
20:01:19.019 -> 
20:01:19.019 -> Protocol=UNKNOWN Hash=0x0 2 bits (incl. gap and start) received
20:01:19.052 -> Received noise or an unknown (or not yet enabled) protocol
20:01:19.052 -> rawData[4]: 
20:01:19.052 ->  -99400
20:01:19.052 ->  +9000,-3000
20:01:19.052 ->  + 600
20:01:19.052 -> Sum: 12600
20:01:19.151 -> 
20:01:19.151 -> Protocol=UNKNOWN Hash=0x0 2 bits (incl. gap and start) received
20:01:19.151 -> Received noise or an unknown (or not yet enabled) protocol
20:01:19.151 -> rawData[4]: 
20:01:19.151 ->  -99400
20:01:19.151 ->  +9000,-3000
20:01:19.151 ->  + 600

*********

At first, I thought it looked the address and command bits were not'ed:
     0x04 ->0xFC

but
   0x02 does not map to 0xFA

so there goes that theory.

Is there some flag I need to throw to get the bits sent out correctly?

Thanks for any help.

cytan

**Comment by ArminJo** [comment]
Please post the timing dump of the received protocol, using debug pin of RecevieDemo.ino.
The timing seems to be totally broken.
Check the distance to receiver and power of the sending LED.

**Reply by cytan299** [reply]
Hi Armin,

Here's the timing dump from my ESP32:

*********
Protocol=Onkyo Address=0xFBFC Command=0xFFFA Raw-Data=0xFFFAFBFC 32 bits LSB first
Send with: IrSender.sendOnkyo(0xFBFC, 0xFFFA, \<numberOfRepeats\>);
rawData[68]: 
 -3276750
 +9750,-4450
 +1550,- 550 +1450,- 550 +1450,-2000 + 550,-1450
 + 550,-1450 + 550,-1450 + 550,-1450 + 550,-1450
 + 550,-1700 + 550,-1650 +1550,- 550 +1450,-1950
 + 600,-1700 + 550,-1650 +1550,-1950 + 600,-1700
 + 550,-1150 + 600,-1700 + 550,-1150 + 600,-1400
 + 650,-1350 + 600,-1400 + 600,-1400 + 600,-1450
 + 600,-1650 + 550,-1200 + 600,-1650 + 550,-1700
 +1500,-2000 + 550,-1650 + 600,-1700 +1500,-2000
 + 550
Sum: 88300

Protocol=UNKNOWN Hash=0x0 2 bits (incl. gap and start) received
Received noise or an unknown (or not yet enabled) protocol
rawData[4]: 
 -98500
 +9000,-2950
 + 600
Sum: 12550

Protocol=UNKNOWN Hash=0x0 2 bits (incl. gap and start) received
Received noise or an unknown (or not yet enabled) protocol
rawData[4]: 
 -98450
 +9000,-2950
 + 600
Sum: 12550
************
For comparison, this is the IR code from the TV control:
************

Protocol=NEC Address=0x4 Command=0x2 Raw-Data=0xFD02FB04 32 bits LSB first
Send with: IrSender.sendNEC(0x4, 0x2, \<numberOfRepeats\>);
rawData[68]: 
 -3276750
 +9000,-4400
 + 600,- 550 + 550,- 550 + 600,-1600 + 600,- 550
 + 600,- 500 + 600,- 500 + 600,- 500 + 600,- 550
 + 600,-1600 + 600,-1650 + 600,- 500 + 600,-1650
 + 600,-1600 + 600,-1650 + 600,-1600 + 600,-1650
 + 600,- 500 + 600,-1650 + 600,- 500 + 600,- 500
 + 600,- 550 + 600,- 500 + 600,- 500 + 600,- 550
 + 600,-1600 + 600,- 500 + 600,-1650 + 600,-1600
 + 600,-1650 + 600,-1650 + 600,-1600 + 600,-1650
 + 600
Sum: 67500

*************

I hope the above helps you figure out what is wrong with my setup.
Thanks!

cytan

**Reply by ArminJo** [reply]
Hi,
do you use wifi on your sending board?

> +1550,- 550 +1450,- 550 +1450,-2000 + 550,-1450

Your send timing is broken. Maybe the ESP32-S3 core is not compatible with this library.

The mark() function relies on the correct implementation of:
delayMicroseconds() for pulse time, and micros() for pause time.
The delayMicroseconds() of pulse time is guarded with noInterrupts() / interrupts().
At the start of pause time, interrupts are enabled once, the rest of the pause is also guarded with noInterrupts() / interrupts().
The maximum length of an interrupt during sending should not exceed 26 us - 8 us = 18 us, otherwise timing is disturbed.
This disturbance is no problem, if the exceedance is small and does not happen too often.

**Reply by cytan299** [reply]
Yes, I'm using Wifi with this project. I can comment out the Wifi to see if the timing improves. I'll do that tonight. But I am not running any Wifi server or looping to check for packets during this test. But I'll comment the Wifi code out for the test.

However, my project requires Wifi and so if IRremote works without Wifi, then it is not that useful for me. So, is there a timing parameter I can tweak to make IRremote work if that's the problem?



cytan

**Reply by cytan299** [reply]
Hi Armin,
   I've commented out my WIFI code and tried again. This is the result, which is a little bit different than before, but obviously incorrect:

******

Protocol=Onkyo Address=0xFBFC Command=0xFDFA Raw-Data=0xFDFAFBFC 32 bits LSB first
Send with: IrSender.sendOnkyo(0xFBFC, 0xFDFA, \<numberOfRepeats\>);
rawData[68]: 
 -3276750
 +9800,-4450
 +1500,- 550 +1500,- 500 +1500,-1950 + 600,-1400
 + 600,-1400 + 600,-1400 + 600,-1400 + 600,-1400
 + 600,-1650 + 550,-1700 +1500,- 550 +1450,-2000
 + 600,-1650 + 600,-1650 +1500,-2000 + 600,-1650
 + 600,-1150 + 600,-1650 + 600,-1150 + 550,-1450
 + 550,-1450 + 550,-1450 + 550,-1450 + 550,-1450
 + 550,-1700 + 600,-1150 + 600,-1650 + 600,-1650
 +1500,-2000 + 550,-1700 + 600,-1650 +1550,-1950
 + 600
Sum: 88350

Protocol=UNKNOWN Hash=0x0 2 bits (incl. gap and start) received
Received noise or an unknown (or not yet enabled) protocol
rawData[4]: 
 -98450
 +9000,-3000
 + 600
Sum: 12600

Protocol=UNKNOWN Hash=0x0 2 bits (incl. gap and start) received
Received noise or an unknown (or not yet enabled) protocol
rawData[4]: 
 -98400
 +9000,-3000
 + 600
Sum: 12600
************

So, is there a way to tune the timing?

cytan

**Reply by ArminJo** [reply]
How will you tune random timing errors?

**Reply by cytan299** [reply]
The output looks random may be due to some type of clock slippage, i.e. the frequency is not quite right between the sender and the receiver causing synchronization problems. Perhaps a tweak in the mark and space widths, and the clock divider will fix the problem. See below.

In the mean time, I decided to try the RMT module (remote control transceiver) that is already built into the ESP32: 

https://docs.espressif.com/projects/esp-idf/en/v4.4/esp32/api-reference/peripherals/rmt.html

and actually got my ESP32 to spit out the correct code.

Raw data sent:
**********
uint16_t rawData[67]={9000,4400,
		      600, 550,  550, 550,  600,1600, 600, 550,
		      600, 500,  600, 500,  600, 500, 600, 550,
		      600,1600, 600,1650,  600, 500,  600,1650,
		      600,1600,  600,1650,  600,1600,  600,1650,
		      600, 500,  600,1650,  600, 500,  600, 500,
		      600, 550,  600, 500,  600, 500,  600, 550,
		      600,1600,  600, 500,  600,1650,  600,1600,
		      600,1650,  600,1650,  600,1600,  600,1650,
		       600};
**********
which I then tweaked in the RMT settings:

```

void buildItem(rmt_item32_t &item,int high_us,int low_us){
  item.level0 = 1;
  item.duration0 = high_us - 50; // remove 50 us
  item.level1 = 0;
  item.duration1 =  low_us + 50; // add 50 us
}
```

and then I got nearly perfect reception:

Received:
********
Protocol=NEC Address=0x4 Command=0x2 Raw-Data=0xFD02FB04 32 bits LSB first
Send with: IrSender.sendNEC(0x4, 0x2, \<numberOfRepeats\>);
rawData[68]: 
 -3276750
 +9000,-4400
 + 600,- 550 + 550,- 550 + 600,-1600 + 600,- 550
 + 600,- 500 + 600,- 500 + 600,- 500 + 600,- 550
 + 600,-1600 + 600,-1650 + 550,- 550 + 600,-1650
 + 600,-1600 + 600,-1650 + 600,-1600 + 600,-1650
 + 600,- 500 + 600,-1650 + 600,- 500 + 600,- 500
 + 600,- 550 + 600,- 500 + 600,- 500 + 600,- 550
 + 600,-1600 + 600,- 500 + 650,-1600 + 550,-1650
 + 600,-1650 + 600,-1650 + 600,-1600 + 600,-1650
 + 600
Sum: 67500
***********

BUT I had to change the clock divider to 40 (I started with 80) to get the above to work! This means that the clock is at 40 MHz???
Somewhere in the compile, there is a message that said this:

```
esptool.py v4.2.1
Serial port /dev/cu.usbmodem14201
Connecting...
Chip is ESP32-S3
Features: WiFi, BLE
Crystal is 40MHz

```
Where does this 40 MHz come from and is it affecting the timings?

Anyway, food for thought. I'll be monkeying with it some more to understand what's going on. And, it still doesn't change the volume of my TV yet, because I'll have to figure out the NEC repeat waveform for the RMT.

cytan

P.S. I'd be glad to share the RMT code with you, when I'm happy with the result.

**Reply by ArminJo** [reply]
Wow 😲

**Comment by cytan299** [comment]
This may be a long shot, but the esp32-s3 runs at 240 MHz. Is there something that needs to be changed because of this clock frequency?

cytan

**Comment by AnalysIR** [comment]
ESP clock speed can be selected in the tools menu of the Arduino IDE...you may have selected 40Mhz before

**Reply by cytan299** [reply]
It looks like one of the Adafruit libraries, I think it is the NeoPixel library (I am not flashing the NeoPixel when I use the RMT library) which also uses the RMT feature of the ESP32, is setting this to the XTAL clock, which is 40 MHz. My solution is for me to set it back to the 80 MHz clock before I send out the pulses.

Unfortunately, it looks like there is some power problems with my IR LED, sometimes it changes the volume of my TV, sometimes it doesn't although IRremote::ReceiveDemo seems to see the correct codes + 1 repeat. Have you ever seen this type of problem before?

cytan

**Reply by cytan299** [reply]
It looks the intermittent problem (which I'm only about 1 m away from the TV) was that simply driving the IR diode with  a current limiting resistor (150 ohms) just doesn't give me enough range. Then I tried a transistor circuit recommended by the ESP8266 authors: 
    https://github.com/crankyoldgit/IRremoteESP8266/wiki#ir-sending
Doing it this way still didn't give enough range.

Finally with the constant current circuit 
  https://www.analysir.com/blog/2013/11/22/constant-current-infrared-led-circuit/
I was able to get the range and I am now able to get the TV to see the IR signal when I'm only about 1 m away. Very strange.

cytan

**Comment by cytan299** [comment]
As promised here's the code using the onboard ESP32-S3 RMT for sending NEC codes (IR_TX.cpp and IR_TX.hpp)

https://github.com/cytan299/Ultrasonic_Screwdriver/tree/main/ultrasonic_screwdriver

I only did the transmit part because that was what I needed. 

Hopefully, it'll help with your code for the ESP32.

cytan

**Reply by ArminJo** [reply]
Thanks for contributing.
Normally *.hpp are *cpp code sources which must be included to enable evaluation of macros, set before the include (especially required for Arduino!).
Your IR_TX.hpp looks more than an old plain IR_TX.h file.

**Reply by cytan299** [reply]
Hi Armin,
   I'm actually not sure what you mean because the preprocessor will run before the compiler is run, so the source code will be unmangled for the compiler with whatever preprocessor commands that were embedded in the code.

   In my style of writing, I always use *.hpp as headers for C++ sources. *.h should work as well. IR_TX.hpp is just simply included in the ino file as a declaration of the IR_TX class. For example, in _ultrasonic_screwdriver.ino_, I have this:

```
.
.
.

/*
  The IR diode
*/

#include "IR_TX.hpp"

IR_TX ir_tx(RMT_CHANNEL_0, GPIO_NUM_12);

.
.
.

```

Anyway, I hope that my code is useful for your ESP32 implementation.

cytan

**Comment by vednildip** [comment]
On  Arduino IDE  environment receiver code  :-  ESP32 S3 Liligo  T HMI   (Used Pin  18  receiver)  -  when  Remote  operated   Guru Meditation  - Reboot   error

**Reply by cytan299** [reply]
Sorry, hard to know why it's Guru meditation. Unfortunately, I've given away my sonic screwdriver and with it the ESP32-S3 so I cannot test.

cytan
